framework/auth/cas.py

```diff
@@ -280,24 +279,57 @@ def get_user_from_cas_resp(cas_resp):
     Given a CAS service validation response, attempt to retrieve user information and next action.
 
     :param cas_resp: the cas service validation response
-    :return: the user and the next action
+    :return: the user, the external_credential, and the next action
     """
 
-    # cas returns the OSF user id
     if cas_resp.user:
         user = User.load(cas_resp.user)
+        # cas returns a valid OSF user id
         if user:
-            return user, 'authenticate'
-
-    # cas returns the external credential
-    if cas_resp.attributes['oauthProvider'] and cas_resp.attributes['oauthId']:
-        user = get_user(external_id_provider=cas_resp.attributes['oauthProvider'], external_id=cas_resp.attributes['oauthId'])
-        # user found
-        if user:
-            # TODO: check if external identity is verified
-            return user, 'authenticate'
-        # user first time login through oauth
+            return user, None, 'authenticate'
+        # cas does not returns a valid OSF user id
         else:
-            return None, 'external_first_login'
+            external_credential = validate_external_credential(cas_resp.user)
+            # invalid cas response
+            if not external_credential:
+                return None, None, None
+            # cas returns an valid external credential
+            user = get_user(external_id_provider=external_credential['provider'],
+                            external_id=external_credential['id'])
+            # user found
+            if user:
+                return user, external_credential, 'authenticate'
+            # user first time login through external identity provider
+            else:
+                return None, external_credential, 'external_first_login'
+
+
+def validate_external_credential(external_credential):
+    """
+    Validate the external credential, a string which is composed of the profile name and the technical identifier
+    of the external provider, separated by `#`. Return the provider and id on success.
+
+    :param external_credential: the external credential string
+    :return: provider and id
+
+    """
+    # wrong format
+    if not external_credential or '#' not in external_credential:
+        return False
+
+    profile_name, technical_id = external_credential.rsplit('#', 1)
```

Comment:
This should be using normal `split` -- it's trivially likely, but more probable that some provider would include an `#` in the `id` than in their service name:
```
In [8]: 'OCRID#1234-5678#9012-3456'.split('#', 1)
Out[8]: ['OCRID', '1234-5678#9012-3456']

In [9]: 'OCRID#1234-5678#9012-3456'.rsplit('#', 1)
Out[9]: ['OCRID#1234-5678', '9012-3456']
```